modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java

```diff
@@ -2141,12 +2141,104 @@ public int compare(CodegenProperty one, CodegenProperty another) {
+                return true;
+            }
+        }
+        if (ModelUtils.isComposedSchema(refSchema)) {
+            ComposedSchema composedSchema = (ComposedSchema) refSchema;
+            if (composedSchema.getAllOf() != null) {
+                // If our discriminator is in one of the allOf schemas break when we find it
+                for  (Schema allOf: composedSchema.getAllOf()) {
+                    if (discriminatorFound(allOf, discPropName)) {
+                        return true;
+                    }
+                }
+            }
+            if (composedSchema.getOneOf() != null && composedSchema.getOneOf().size() != 0) {
+                // All oneOf definitions must contain the discriminator
+                Integer hasDiscriminatorCnt = 0;
+                for  (Schema oneOf: composedSchema.getOneOf()) {
+                    if (discriminatorFound(oneOf, discPropName)) {
+                        hasDiscriminatorCnt++;
+                    }
+                }
+                if (hasDiscriminatorCnt == composedSchema.getOneOf().size()) {
+                    return true;
+                }
+            }
+            if (composedSchema.getAnyOf() != null && composedSchema.getAnyOf().size() != 0) {
+                // All anyOf definitions must contain the discriminator because a min of one must be selected
+                Integer hasDiscriminatorCnt = 0;
+                for  (Schema anyOf: composedSchema.getAnyOf()) {
+                    if (discriminatorFound(anyOf, discPropName)) {
+                        hasDiscriminatorCnt++;
+                    }
+                }
+                if (hasDiscriminatorCnt == composedSchema.getAnyOf().size()) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    /**
+     * This function is only used for composed schemas which have a discriminator
+     * Process oneOf and anyOf models in a composed schema and adds them into
+     * their parent's discriminator map if the oneOf and anyOf models contain
+     * the required discriminator. If they don't contain the required
+     * discriminator or the discriminator is the wrong type then an error is
+     * thrown
+     * @param composedSchemaName The String model name of the composed schema where we are setting the discriminator map
+     * @param discriminator The CodegenDiscriminator that we will add models to
+     * @param c The ComposedSchema that contains the discriminator and oneOf/anyOf schemas
+     * @param discPropName The String that is the discriminator propertyName in the schema
+     */
+    private void addComposedMappedModels(String composedSchemaName, CodegenDiscriminator discriminator, ComposedSchema c, String discPropName) {
+        ArrayList<List<Schema>> listOLists = new ArrayList<List<Schema>>();
+        listOLists.add(c.getOneOf());
+        listOLists.add(c.getAnyOf());
```

Comment:
Yes we should per:
https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.0.md#discriminatorObject
```
When used in conjunction with the anyOf construct, the use of the discriminator can avoid ambiguity where multiple schemas may satisfy a single payload.
```
So the spec wants us to allow anyOf to allow a discriminator map so it may act like oneOf